[Bug]: possible data race in AvailableArchitecture.GetHash() function

### Is there an existing issue already for this bug?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new bug.

### I have read the troubleshooting guide

- [X] I have read the troubleshooting guide and I think this is a new bug.

### I am running a supported version of CloudNativePG

- [X] I have read the troubleshooting guide and I think this is a new bug.

### Contact Details

_No response_

### Version

1.24.0

### What version of Kubernetes are you using?

1.31

### What is your Kubernetes environment?

Self-managed: kind (evaluation)

### How did you install the operator?

YAML manifest

### What happened?

There is a possible data race at

https://github.com/cloudnative-pg/cloudnative-pg/blob/90478f1b50224155b4c83bd76eb0bf32f85f59ee/pkg/utils/discovery.go#L63-L70

where the `arch.hash` value is accessed outside the protection of the mutex.

### Cluster resource

_No response_

### Relevant log output

```shell
ginkgo -repeat=100 -race pkg/utils/...
[1725279695] utils test suite - 99/99 specs ••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••==================
WARNING: DATA RACE
Write at 0x00c000485c50 by goroutine 129:
  github.com/cloudnative-pg/cloudnative-pg/pkg/utils.(*AvailableArchitecture).calculateHash()
      /Users/mnencia/prj/cnpg/pkg/utils/discovery.go:86 +0x171
  github.com/cloudnative-pg/cloudnative-pg/pkg/utils.detectAvailableArchitectures.gowrap1()
      /Users/mnencia/prj/cnpg/pkg/utils/discovery.go:224 +0x33

Previous read at 0x00c000485c50 by goroutine 128:
  github.com/cloudnative-pg/cloudnative-pg/pkg/utils.(*AvailableArchitecture).GetHash()
      /Users/mnencia/prj/cnpg/pkg/utils/discovery.go:64 +0x47
  github.com/cloudnative-pg/cloudnative-pg/pkg/utils.init.func6.4.4.1()
      /Users/mnencia/prj/cnpg/pkg/utils/discovery_test.go:243 +0xeb
  runtime.call16()
      /Users/mnencia/.asdf/installs/golang/1.23.0/go/src/runtime/asm_amd64.s:775 +0x42
  reflect.Value.Call()
      /Users/mnencia/.asdf/installs/golang/1.23.0/go/src/reflect/value.go:365 +0xb5
  github.com/onsi/gomega/internal.(*AsyncAssertion).buildActualPoller.func3()
      /Users/mnencia/.asdf/installs/golang/1.23.0/packages/pkg/mod/github.com/onsi/gomega@v1.34.1/internal/async_assertion.go:325 +0x249
  github.com/onsi/gomega/internal.(*AsyncAssertion).match()
      /Users/mnencia/.asdf/installs/golang/1.23.0/packages/pkg/mod/github.com/onsi/gomega@v1.34.1/internal/async_assertion.go:398 +0x2a7
  github.com/onsi/gomega/internal.(*AsyncAssertion).Should()
      /Users/mnencia/.asdf/installs/golang/1.23.0/packages/pkg/mod/github.com/onsi/gomega@v1.34.1/internal/async_assertion.go:145 +0xc7
  github.com/cloudnative-pg/cloudnative-pg/pkg/utils.init.func6.4.4()
      /Users/mnencia/prj/cnpg/pkg/utils/discovery_test.go:245 +0x651
  github.com/onsi/ginkgo/v2/internal.extractBodyFunction.func3()
      /Users/mnencia/.asdf/installs/golang/1.23.0/packages/pkg/mod/github.com/onsi/ginkgo/v2@v2.20.0/internal/node.go:472 +0x2e
  github.com/onsi/ginkgo/v2/internal.(*Suite).runNode.func3()
      /Users/mnencia/.asdf/installs/golang/1.23.0/packages/pkg/mod/github.com/onsi/ginkgo/v2@v2.20.0/internal/suite.go:894 +0x12b

Goroutine 129 (running) created at:
  github.com/cloudnative-pg/cloudnative-pg/pkg/utils.detectAvailableArchitectures()
      /Users/mnencia/prj/cnpg/pkg/utils/discovery.go:224 +0x84
  github.com/cloudnative-pg/cloudnative-pg/pkg/utils.init.func6.4.4()
      /Users/mnencia/prj/cnpg/pkg/utils/discovery_test.go:234 +0x449
  github.com/onsi/ginkgo/v2/internal.extractBodyFunction.func3()
      /Users/mnencia/.asdf/installs/golang/1.23.0/packages/pkg/mod/github.com/onsi/ginkgo/v2@v2.20.0/internal/node.go:472 +0x2e
  github.com/onsi/ginkgo/v2/internal.(*Suite).runNode.func3()
      /Users/mnencia/.asdf/installs/golang/1.23.0/packages/pkg/mod/github.com/onsi/ginkgo/v2@v2.20.0/internal/suite.go:894 +0x12b

Goroutine 128 (running) created at:
  github.com/onsi/ginkgo/v2/internal.(*Suite).runNode()
      /Users/mnencia/.asdf/installs/golang/1.23.0/packages/pkg/mod/github.com/onsi/ginkgo/v2@v2.20.0/internal/suite.go:881 +0x1516
  github.com/onsi/ginkgo/v2/internal.(*group).attemptSpec()
      /Users/mnencia/.asdf/installs/golang/1.23.0/packages/pkg/mod/github.com/onsi/ginkgo/v2@v2.20.0/internal/group.go:199 +0x1164
  github.com/onsi/ginkgo/v2/internal.(*group).run()
      /Users/mnencia/.asdf/installs/golang/1.23.0/packages/pkg/mod/github.com/onsi/ginkgo/v2@v2.20.0/internal/group.go:349 +0x1207
  github.com/onsi/ginkgo/v2/internal.(*Suite).runSpecs()
      /Users/mnencia/.asdf/installs/golang/1.23.0/packages/pkg/mod/github.com/onsi/ginkgo/v2@v2.20.0/internal/suite.go:489 +0x1291
  github.com/onsi/ginkgo/v2/internal.(*Suite).Run()
      /Users/mnencia/.asdf/installs/golang/1.23.0/packages/pkg/mod/github.com/onsi/ginkgo/v2@v2.20.0/internal/suite.go:130 +0x5f7
  github.com/onsi/ginkgo/v2.RunSpecs()
      /Users/mnencia/.asdf/installs/golang/1.23.0/packages/pkg/mod/github.com/onsi/ginkgo/v2@v2.20.0/core_dsl.go:300 +0xeaa
  github.com/cloudnative-pg/cloudnative-pg/pkg/utils.TestUtils()
      /Users/mnencia/prj/cnpg/pkg/utils/suite_test.go:28 +0x4d
  testing.tRunner()
      /Users/mnencia/.asdf/installs/golang/1.23.0/go/src/testing/testing.go:1690 +0x226
  testing.(*T).Run.gowrap1()
      /Users/mnencia/.asdf/installs/golang/1.23.0/go/src/testing/testing.go:1743 +0x44
==================
••••••••••••••••••••••••••••••••••••••• SUCCESS! 33.519036ms --- FAIL: TestUtils (0.04s)
    testing.go:1399: race detected during execution of test
FAIL

Ginkgo ran 3 suites in 13.352034594s

There were failures detected in the following suites:
  utils ./pkg/utils

Test Suite Failed
```


### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/pkg/utils/discovery.go
+++ b/pkg/utils/discovery.go
@@ -61,21 +61,16 @@ func newAvailableArchitecture(goArch, binaryPath string) *AvailableArchitecture 
 
 // GetHash retrieves the hash for a given AvailableArchitecture
 func (arch *AvailableArchitecture) GetHash() string {
-	if arch.hash == "" {
-		arch.calculateHash()
-	}
-	arch.mx.Lock()
-	defer arch.mx.Unlock()
-	return arch.hash
+	return arch.calculateHash()
 }
 
 // calculateHash calculates the hash for a given AvailableArchitecture
-func (arch *AvailableArchitecture) calculateHash() {
+func (arch *AvailableArchitecture) calculateHash() string {
 	arch.mx.Lock()
 	defer arch.mx.Unlock()
 
 	if arch.hash != "" {
-		return
+		return arch.hash
 	}
 
 	hash, err := arch.hashCalculator(arch.binaryPath)
@@ -84,6 +79,7 @@ func (arch *AvailableArchitecture) calculateHash() {
 	}
 
 	arch.hash = hash
+	return hash
 }
 
 // FileStream opens a stream reading from the manager's binary

--- a/pkg/utils/discovery_test.go
+++ b/pkg/utils/discovery_test.go
@@ -192,7 +192,9 @@ var _ = Describe("AvailableArchitecture", func() {
 			})
 
 			It("should panic with an error", func() {
-				Expect(arch.calculateHash).To(Panic())
+				Expect(func() {
+					arch.calculateHash()
+				}).To(Panic())
 			})
 		})
 	})

